alibi_detect/cd/base_online.py

```diff
@@ -106,6 +113,66 @@ def _configure_ref_subset(self):
     def _update_state(self, x_t: Union[np.ndarray, 'tf.Tensor', 'torch.Tensor']):
         pass
 
+    def _set_state_dir(self, dirpath: Union[str, os.PathLike]):
+        """
+        Set the directory path to store state in, and create an empty directory if it doesn't already exist.
+
+        Parameters
+        ----------
+        dirpath
+            The directory to save state file inside.
+        """
+        self.state_dir = Path(dirpath)
+        self.state_dir.mkdir(parents=True, exist_ok=True)
+
+    def save_state(self, filepath: Union[str, os.PathLike]):
+        """
+        Save a detector's state to disk in order to generate a checkpoint.
+
+        Parameters
+        ----------
+        filepath
+            The directory to save state to.
+        """
+        self._set_state_dir(filepath)
+        self._save_state()
+
+    def load_state(self, filepath: Union[str, os.PathLike]):
+        """
+        Load the detector's state from disk, in order to restart from a checkpoint previously generated with
+        `save_state`.
+
+        Parameters
+        ----------
+        filepath
+            The directory to load state from.
+        """
+        self._set_state_dir(filepath)
```

Comment:
Why is this necessary when loading?

Reply:
Just so that `self.state_dir` is set (and converted from `str` to `pathlib.Path) when `load_state` is called as well as when `save_state` is called. 

I thought it might be helpful to have `state_dir` as a public attribute so that a user could see interrogate the detector to see where state was loaded from. Although thinking about it more, for the backend detectors one would have to do `detector._detector.state_dir` (access a private attribute) anyway. I guess we'd probably want to define a `@property` if we actually want to support this functionality properly...

Happy to just make it private if you think its better though...